packages/playwright-crawler/src/internals/adaptive-playwright-crawler.ts

```diff
@@ -0,0 +1,263 @@
+import { addTimeoutToPromise } from '@apify/timeout';
+import { BasicCrawler } from '@crawlee/basic';
+import type { Source, Configuration, RecordOptions, RestrictedCrawlingContext } from '@crawlee/core';
+import { KeyValueStore } from '@crawlee/core';
+import type { Awaitable, Dictionary } from '@crawlee/types';
+import { extractUrlsFromCheerio } from '@crawlee/utils';
+import type { Cheerio, Document } from 'cheerio';
+import { load } from 'cheerio';
+import { cheerioPortadom, playwrightLocatorPortadom, type CheerioPortadom, type PlaywrightLocatorPortadom } from 'portadom';
+
+import type { PlaywrightCrawlerOptions, PlaywrightCrawlingContext } from './playwright-crawler';
+import { PlaywrightCrawler } from './playwright-crawler';
+import { RenderingTypePredictor, type RenderingType } from './rendering-type-prediction';
+
+type Result<TResult> = NonNullable<{result: TResult; ok: true} | {error: unknown; ok: false}>
+
+class RequestHandlerResult {
+    datasetItems: {item: Dictionary; datasetIdOrName?: string}[] = [];
+    requestListUrls: {url?: string; label?: string; request: Source; options: Parameters<RestrictedCrawlingContext['addRequests']>[1]}[] = [];
+    requestUrls: {url?: string; label?: string; request: Source; options: Parameters<RestrictedCrawlingContext['addRequests']>[1]}[] = [];
+    enqueuedUrls: {url?: string; label?: string; options: Parameters<RestrictedCrawlingContext['enqueueLinks']>[0]}[] = [];
+    keyValueStoreChanges: Record<string, Record<string, {changedValue: unknown; options?: RecordOptions}>> = {};
```

Comment:
I hate to make this public. Ideally, the class should have a method that returns this as readonly. Do we have the types for that in place? Or a favorite type utility library that I could use here?

Reply:
It's not hard to make a `DeepReadonly` type helper. However, the resulting type cannot be passed to `enqueueLinks` for example. I'd have to update the crawling context type for that, which seems invasive. Thoughts?